extension/opampextension/opamp_agent.go

```diff
@@ -0,0 +1,244 @@
+	o.opampClient = client.NewWebSocket(o.logger.Sugar())
+
+	header := http.Header{}
+	for k, v := range o.cfg.Headers {
+		header.Set(k, string(v))
+	}
+
+	tls := &tls.Config{
+		ServerName:         o.cfg.TLSSetting.ServerName,
+		InsecureSkipVerify: o.cfg.TLSSetting.InsecureSkipVerify,
+	}
+
+	settings := types.StartSettings{
+		Header:         header,
+		TLSConfig:      tls,
+		OpAMPServerURL: o.cfg.Endpoint,
+		InstanceUid:    o.instanceId.String(),
+		Callbacks: types.CallbacksStruct{
+			OnConnectFunc: func() {
+				o.logger.Debug("Connected to the OpAMP server")
+			},
+			OnConnectFailedFunc: func(err error) {
+				o.logger.Error("Failed to connect to the OpAMP server", zap.Error(err))
+			},
+			OnErrorFunc: func(err *protobufs.ServerErrorResponse) {
+				o.logger.Error("OpAMP server returned an error response", zap.String("message", err.ErrorMessage))
+			},
+			GetEffectiveConfigFunc: func(ctx context.Context) (*protobufs.EffectiveConfig, error) {
+				return o.composeEffectiveConfig(), nil
+			},
+			OnMessageFunc: o.onMessage,
+		},
+		// Include ReportsEffectiveConfig once the extension has access to the
+		// collector's effective configuration.
+		Capabilities: protobufs.AgentCapabilities_AgentCapabilities_ReportsStatus,
+	}
+
+	if err := o.createAgentDescription(); err != nil {
+		return err
+	}
+
+	if err := o.opampClient.SetAgentDescription(o.agentDescription); err != nil {
+		return err
+	}
+
+	o.logger.Debug("Starting OpAMP client...")
+
+	if err := o.opampClient.Start(context.Background(), settings); err != nil {
+		return err
+	}
+
+	o.logger.Debug("OpAMP client started")
+
+	return nil
+}
+
+func (o *opampAgent) Shutdown(ctx context.Context) error {
+	o.logger.Debug("OpAMP agent shutting down...")
+	if o.opampClient != nil {
+		o.logger.Debug("Stopping OpAMP client...")
+		err := o.opampClient.Stop(ctx)
+		return err
+	}
+	return nil
+}
+
+func newOpampAgent(cfg *Config, logger *zap.Logger, build component.BuildInfo, res pcommon.Resource) (*opampAgent, error) {
+	agentType := build.Command
+
+	sn, ok := res.Attributes().Get(semconv.AttributeServiceName)
+	if ok {
+		agentType = sn.AsString()
+	}
+
+	agentVersion := build.Version
+
+	sv, ok := res.Attributes().Get(semconv.AttributeServiceVersion)
+	if ok {
+		agentVersion = sv.AsString()
+	}
+
+	uid := ulid.Make()
+
+	sid, ok := res.Attributes().Get(semconv.AttributeServiceInstanceID)
+	if ok {
+		puid, err := ulid.Parse(sid.AsString())
+		if err != nil {
+			return nil, err
+		}
+		uid = puid
+	}
+
+	if cfg.InstanceUID != "" {
+		puid, err := ulid.Parse(cfg.InstanceUID)
+		if err != nil {
+			return nil, err
+		}
+		uid = puid
+	}
+
+	agent := &opampAgent{
+		cfg:             cfg,
+		logger:          logger,
+		agentType:       agentType,
+		agentVersion:    agentVersion,
+		instanceId:      uid,
+		effectiveConfig: localConfig, // TODO: Replace with https://github.com/open-telemetry/opentelemetry-collector/issues/6596
```

Comment:
Mentioned elsewhere - can we resolve the TODO or update it to reference an open issue?